scipy/optimize/_trustregion_exact.py

```diff
@@ -400,10 +413,10 @@ def solve(self, tr_radius):
                 s_min, z_min = estimate_smallest_singular_value(U)
                 step_len = tr_radius
 
+                p = step_len * z_min
```

Comment:
Did you find a case where the previous code failed and this doesn't?

Reply:
This line is simply move up so that `p` is defined even if the stop criteria check below is never fulfilled (now we may exit the loop when reaching `self.maxiter`). Without moving `p` here, I observed `NameError` when `p` was accessed. (I think it was the return statement, but I'm not sure).

I can see if I can dig out the scripts where that error occurred, and extract a test case. I thought this was needed even for the test in this PR, but if I move this assignment back into the if-statement everything seems to work still.

Reply:
I can't seem to find a failure case for this right now. But if we look at the branches in the loop we see that the assignment needs to happen here instead of inside the if-block in order to avoid the risk of having a `NameError` in the return from the method.

Reply:
This looks reasonable to me. I suppose it would be easiest to craft a test case with `self.maxiter=1`, then you just have to ensure the stop criteria is not hit on the first iteration, right?